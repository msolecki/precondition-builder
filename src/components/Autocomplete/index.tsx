import React, {ChangeEvent} from 'react'
import PropTypes from 'prop-types'
import TextField from '@material-ui/core/TextField'
import MenuItem from '@material-ui/core/MenuItem'
import Paper from '@material-ui/core/Paper'
import Downshift from 'downshift'
import deburr from 'lodash.deburr'
import {makeStyles} from '@material-ui/core'

export interface AutocompleteProps {
    handleChange(value: string): void;

    data: string[];
    selectedData: string | null;
}

const useStyles = makeStyles(({spacing}) => ({
    root: {
        flexGrow: 1,
        height: 250,
    },
    container: {
        flexGrow: 1,
        position: 'relative',
    },
    paper: {
        position: 'absolute',
        zIndex: 1,
        marginTop: spacing(1),
        left: 0,
        right: 0,
    },
    chip: {
        margin: spacing(0.5, 0.25),
    },
    inputRoot: {
        flexWrap: 'wrap',
    },
    inputInput: {
        width: 'auto',
        flexGrow: 1,
    },
    formControlWithoutMargin: {
        marginTop: spacing(1),
        marginBottom: spacing(3),
        minWidth: 120,
    },
}))

const Autocomplete: React.FC<AutocompleteProps> = (props: AutocompleteProps): React.ReactElement => {
    const classes = useStyles()

    function renderInput(inputProps): React.ReactElement {
        const {InputProps, classes, ref, ...other} = inputProps

        return (
            <TextField
                InputProps={{
                    inputRef: ref,
                    classes: {
                        root: classes.inputRoot,
                        input: classes.inputInput,
                    },
                    ...InputProps,
                }}
                {...other}
            />
        )
    }

    function getSuggestions(value): string[] {
        const inputValue = deburr(value).toLowerCase()
        const inputLength = inputValue.length

        return props.data.filter((id): boolean => id.slice(0, inputLength).toLowerCase() === inputValue).slice(0, 5)
    }

    function renderSuggestion(suggestionProps): React.ReactElement {
        const {nuggetId, index, itemProps, highlightedIndex, selectedItem} = suggestionProps
        const isHighlighted = highlightedIndex === index
        const isSelected = (selectedItem || '').indexOf(nuggetId) > -1

        return (
            <MenuItem
                {...itemProps}
                key={nuggetId}
                selected={isHighlighted}
                component="div"
                style={{
                    fontWeight: isSelected ? 500 : 400,
                }}
            >
                {nuggetId}
            </MenuItem>
        )
    }

    function handleChange(changes): void {
        if (changes.hasOwnProperty('inputValue')) {
            props.handleChange(changes.inputValue.toString())
        }
    }

    return (
        <Downshift initialSelectedItem={props.selectedData} onStateChange={handleChange}>
            {({clearSelection, getInputProps, getItemProps, getLabelProps, getMenuProps, highlightedIndex, inputValue, isOpen, openMenu, selectedItem}): React.ReactElement => {
                const onChangeData = (event: ChangeEvent<HTMLInputElement>): void => {
                    if (event.target.value === '') {
                        clearSelection()
                    }
                }

                const {onBlur, onChange, onFocus, ...inputProps} = getInputProps({
                    onChange: onChangeData,
                    onFocus: openMenu,
                    placeholder: 'Start typing',
                })


                const suggestions = getSuggestions(inputValue)

                return (
                    <div className={classes.container}>
                        {renderInput({
                            fullWidth: true,
                            classes,
                            label: 'HTML Nugget ID',
                            InputLabelProps: getLabelProps(),
                            InputProps: {onBlur, onChange, onFocus},
                            inputProps,
                        })}

                        <div {...getMenuProps()}>
                            {isOpen &&
                            <Paper className={classes.paper} square>
                                {suggestions.map((nuggetId, index): React.ReactElement =>
                                    renderSuggestion({
                                        nuggetId,
                                        index,
                                        itemProps: getItemProps({item: nuggetId}),
                                        highlightedIndex,
                                        selectedItem,
                                    }),
                                )}
                            </Paper>
                            }
                        </div>
                    </div>
                )
            }}
        </Downshift>
    )
}

Autocomplete.propTypes = {
    handleChange: PropTypes.func.isRequired,
    data: PropTypes.array.isRequired,
    selectedData: PropTypes.string,
}

Autocomplete.defaultProps = {
    data: [],
    selectedData: null
}

export default Autocomplete
