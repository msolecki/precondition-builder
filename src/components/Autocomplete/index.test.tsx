import * as React from 'react'
import {ReactTestRenderer} from 'react-test-renderer'
import {createMount} from '@material-ui/core/test-utils'

import Autocomplete, {AutocompleteProps} from './index'

const component = (props: AutocompleteProps): Autocomplete => <Autocomplete {...props}/>
const handleChangeFunc = jest.fn()

describe('<Autocomplete />', () => {
    test('empty snapshot', () => {
        const wrapper: ReactTestRenderer = createMount()(component({
            handleChange: () => {
            },
            data: [],
            selectedData: null
        }))
        expect(wrapper).toMatchSnapshot()
    })

    test('default data', () => {
        const wrapper = createMount()(
            component({
                handleChange: handleChangeFunc,
                data: ['1'],
                selectedData: null
            }))
        const event = {target: {name: 'name', value: ''}}

        wrapper.find('input').simulate('change', event)

        expect(handleChangeFunc).not.toHaveBeenCalled()
    })

    test('with selected data', () => {
        const wrapper = createMount()(
            component({
                handleChange: handleChangeFunc,
                data: ['1'],
                selectedData: '1'
            }))
        const event = {target: {name: 'name', value: 1}}

        wrapper.find('input').simulate('change', event)

        expect(handleChangeFunc).toHaveBeenCalledWith('1')
    })

    test('with selected data', () => {
        const wrapper = createMount()(
            component({
                handleChange: handleChangeFunc,
                data: [],
                selectedData: null
            }))
        const event = {target: {name: 'name', value: ''}}

        wrapper.find('input').simulate('change', event)

        expect(handleChangeFunc).toHaveBeenCalledWith('1')
    })
})
