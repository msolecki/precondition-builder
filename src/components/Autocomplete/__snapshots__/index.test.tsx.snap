// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Autocomplete /> empty snapshot 1`] = `
<Autocomplete
  data={Array []}
  handleChange={[Function]}
  selectedData={null}
>
  <Downshift
    defaultHighlightedIndex={null}
    defaultIsOpen={false}
    environment={[Window]}
    getA11yStatusMessage={[Function]}
    initialSelectedItem={null}
    itemToString={[Function]}
    onChange={[Function]}
    onInputValueChange={[Function]}
    onOuterClick={[Function]}
    onSelect={[Function]}
    onStateChange={[Function]}
    onUserAction={[Function]}
    scrollIntoView={[Function]}
    selectedItemChanged={[Function]}
    stateReducer={[Function]}
    suppressRefError={false}
  >
    <div
      aria-expanded={false}
      aria-haspopup="listbox"
      aria-labelledby="downshift-0-label"
      aria-owns={null}
      className="makeStyles-container-2"
      role="combobox"
    >
      <WithStyles(ForwardRef(TextField))
        InputLabelProps={
          Object {
            "htmlFor": "downshift-0-input",
            "id": "downshift-0-label",
          }
        }
        InputProps={
          Object {
            "classes": Object {
              "input": "makeStyles-inputInput-6",
              "root": "makeStyles-inputRoot-5",
            },
            "inputRef": undefined,
            "onBlur": [Function],
            "onChange": [Function],
            "onFocus": [Function],
          }
        }
        fullWidth={true}
        inputProps={
          Object {
            "aria-activedescendant": null,
            "aria-autocomplete": "list",
            "aria-controls": null,
            "aria-labelledby": "downshift-0-label",
            "autoComplete": "off",
            "id": "downshift-0-input",
            "onKeyDown": [Function],
            "placeholder": "Start typing",
            "value": "",
          }
        }
        label="HTML Nugget ID"
      >
        <ForwardRef(TextField)
          InputLabelProps={
            Object {
              "htmlFor": "downshift-0-input",
              "id": "downshift-0-label",
            }
          }
          InputProps={
            Object {
              "classes": Object {
                "input": "makeStyles-inputInput-6",
                "root": "makeStyles-inputRoot-5",
              },
              "inputRef": undefined,
              "onBlur": [Function],
              "onChange": [Function],
              "onFocus": [Function],
            }
          }
          classes={
            Object {
              "root": "MuiTextField-root",
            }
          }
          fullWidth={true}
          inputProps={
            Object {
              "aria-activedescendant": null,
              "aria-autocomplete": "list",
              "aria-controls": null,
              "aria-labelledby": "downshift-0-label",
              "autoComplete": "off",
              "id": "downshift-0-input",
              "onKeyDown": [Function],
              "placeholder": "Start typing",
              "value": "",
            }
          }
          label="HTML Nugget ID"
        >
          <WithStyles(ForwardRef(FormControl))
            className="MuiTextField-root"
            fullWidth={true}
            required={false}
            variant="standard"
          >
            <ForwardRef(FormControl)
              className="MuiTextField-root"
              classes={
                Object {
                  "fullWidth": "MuiFormControl-fullWidth",
                  "marginDense": "MuiFormControl-marginDense",
                  "marginNormal": "MuiFormControl-marginNormal",
                  "root": "MuiFormControl-root",
                }
              }
              fullWidth={true}
              required={false}
              variant="standard"
            >
              <div
                className="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
              >
                <WithStyles(WithFormControlContext(ForwardRef(InputLabel)))
                  htmlFor="downshift-0-input"
                  id="downshift-0-label"
                >
                  <WithFormControlContext(ForwardRef(InputLabel))
                    classes={
                      Object {
                        "animated": "MuiInputLabel-animated",
                        "asterisk": "MuiInputLabel-asterisk",
                        "disabled": "Mui-disabled",
                        "error": "Mui-error",
                        "filled": "MuiInputLabel-filled",
                        "focused": "Mui-focused",
                        "formControl": "MuiInputLabel-formControl",
                        "marginDense": "MuiInputLabel-marginDense",
                        "outlined": "MuiInputLabel-outlined",
                        "required": "Mui-required",
                        "root": "MuiInputLabel-root",
                        "shrink": "MuiInputLabel-shrink",
                      }
                    }
                    htmlFor="downshift-0-input"
                    id="downshift-0-label"
                  >
                    <ForwardRef(InputLabel)
                      classes={
                        Object {
                          "animated": "MuiInputLabel-animated",
                          "asterisk": "MuiInputLabel-asterisk",
                          "disabled": "Mui-disabled",
                          "error": "Mui-error",
                          "filled": "MuiInputLabel-filled",
                          "focused": "Mui-focused",
                          "formControl": "MuiInputLabel-formControl",
                          "marginDense": "MuiInputLabel-marginDense",
                          "outlined": "MuiInputLabel-outlined",
                          "required": "Mui-required",
                          "root": "MuiInputLabel-root",
                          "shrink": "MuiInputLabel-shrink",
                        }
                      }
                      htmlFor="downshift-0-input"
                      id="downshift-0-label"
                      muiFormControl={
                        Object {
                          "adornedStart": false,
                          "disabled": false,
                          "error": false,
                          "filled": false,
                          "focused": false,
                          "margin": "none",
                          "onBlur": [Function],
                          "onEmpty": [Function],
                          "onFilled": [Function],
                          "onFocus": [Function],
                          "required": false,
                          "variant": "standard",
                        }
                      }
                    >
                      <WithStyles(WithFormControlContext(ForwardRef(FormLabel)))
                        className="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
                        classes={
                          Object {
                            "asterisk": "MuiInputLabel-asterisk",
                            "disabled": "Mui-disabled",
                            "error": "Mui-error",
                            "focused": "Mui-focused",
                            "required": "Mui-required",
                          }
                        }
                        data-shrink={false}
                        htmlFor="downshift-0-input"
                        id="downshift-0-label"
                      >
                        <WithFormControlContext(ForwardRef(FormLabel))
                          className="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
                          classes={
                            Object {
                              "asterisk": "MuiFormLabel-asterisk MuiInputLabel-asterisk",
                              "disabled": "Mui-disabled Mui-disabled",
                              "error": "Mui-error Mui-error",
                              "filled": "MuiFormLabel-filled",
                              "focused": "Mui-focused Mui-focused",
                              "required": "Mui-required Mui-required",
                              "root": "MuiFormLabel-root",
                            }
                          }
                          data-shrink={false}
                          htmlFor="downshift-0-input"
                          id="downshift-0-label"
                        >
                          <ForwardRef(FormLabel)
                            className="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
                            classes={
                              Object {
                                "asterisk": "MuiFormLabel-asterisk MuiInputLabel-asterisk",
                                "disabled": "Mui-disabled Mui-disabled",
                                "error": "Mui-error Mui-error",
                                "filled": "MuiFormLabel-filled",
                                "focused": "Mui-focused Mui-focused",
                                "required": "Mui-required Mui-required",
                                "root": "MuiFormLabel-root",
                              }
                            }
                            data-shrink={false}
                            htmlFor="downshift-0-input"
                            id="downshift-0-label"
                            muiFormControl={
                              Object {
                                "adornedStart": false,
                                "disabled": false,
                                "error": false,
                                "filled": false,
                                "focused": false,
                                "margin": "none",
                                "onBlur": [Function],
                                "onEmpty": [Function],
                                "onFilled": [Function],
                                "onFocus": [Function],
                                "required": false,
                                "variant": "standard",
                              }
                            }
                          >
                            <label
                              className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
                              data-shrink={false}
                              htmlFor="downshift-0-input"
                              id="downshift-0-label"
                            >
                              HTML Nugget ID
                            </label>
                          </ForwardRef(FormLabel)>
                        </WithFormControlContext(ForwardRef(FormLabel))>
                      </WithStyles(WithFormControlContext(ForwardRef(FormLabel)))>
                    </ForwardRef(InputLabel)>
                  </WithFormControlContext(ForwardRef(InputLabel))>
                </WithStyles(WithFormControlContext(ForwardRef(InputLabel)))>
                <WithStyles(ForwardRef(Input))
                  classes={
                    Object {
                      "input": "makeStyles-inputInput-6",
                      "root": "makeStyles-inputRoot-5",
                    }
                  }
                  fullWidth={true}
                  inputProps={
                    Object {
                      "aria-activedescendant": null,
                      "aria-autocomplete": "list",
                      "aria-controls": null,
                      "aria-labelledby": "downshift-0-label",
                      "autoComplete": "off",
                      "id": "downshift-0-input",
                      "onKeyDown": [Function],
                      "placeholder": "Start typing",
                      "value": "",
                    }
                  }
                  onBlur={[Function]}
                  onChange={[Function]}
                  onFocus={[Function]}
                >
                  <ForwardRef(Input)
                    classes={
                      Object {
                        "disabled": "Mui-disabled",
                        "error": "Mui-error",
                        "focused": "Mui-focused",
                        "formControl": "MuiInput-formControl",
                        "fullWidth": "MuiInput-fullWidth",
                        "input": "MuiInput-input makeStyles-inputInput-6",
                        "inputMarginDense": "MuiInput-inputMarginDense",
                        "inputMultiline": "MuiInput-inputMultiline",
                        "inputTypeSearch": "MuiInput-inputTypeSearch",
                        "multiline": "MuiInput-multiline",
                        "root": "MuiInput-root makeStyles-inputRoot-5",
                        "underline": "MuiInput-underline",
                      }
                    }
                    fullWidth={true}
                    inputProps={
                      Object {
                        "aria-activedescendant": null,
                        "aria-autocomplete": "list",
                        "aria-controls": null,
                        "aria-labelledby": "downshift-0-label",
                        "autoComplete": "off",
                        "id": "downshift-0-input",
                        "onKeyDown": [Function],
                        "placeholder": "Start typing",
                        "value": "",
                      }
                    }
                    onBlur={[Function]}
                    onChange={[Function]}
                    onFocus={[Function]}
                  >
                    <WithStyles(ForwardRef(InputBase))
                      classes={
                        Object {
                          "disabled": "Mui-disabled",
                          "error": "Mui-error",
                          "focused": "Mui-focused",
                          "formControl": "MuiInput-formControl",
                          "fullWidth": "MuiInput-fullWidth",
                          "input": "MuiInput-input makeStyles-inputInput-6",
                          "inputMarginDense": "MuiInput-inputMarginDense",
                          "inputMultiline": "MuiInput-inputMultiline",
                          "inputTypeSearch": "MuiInput-inputTypeSearch",
                          "multiline": "MuiInput-multiline",
                          "root": "MuiInput-root makeStyles-inputRoot-5 MuiInput-underline",
                          "underline": null,
                        }
                      }
                      fullWidth={true}
                      inputComponent="input"
                      inputProps={
                        Object {
                          "aria-activedescendant": null,
                          "aria-autocomplete": "list",
                          "aria-controls": null,
                          "aria-labelledby": "downshift-0-label",
                          "autoComplete": "off",
                          "id": "downshift-0-input",
                          "onKeyDown": [Function],
                          "placeholder": "Start typing",
                          "value": "",
                        }
                      }
                      multiline={false}
                      onBlur={[Function]}
                      onChange={[Function]}
                      onFocus={[Function]}
                      type="text"
                    >
                      <ForwardRef(InputBase)
                        classes={
                          Object {
                            "adornedEnd": "MuiInputBase-adornedEnd",
                            "adornedStart": "MuiInputBase-adornedStart",
                            "disabled": "Mui-disabled Mui-disabled",
                            "error": "Mui-error Mui-error",
                            "focused": "Mui-focused Mui-focused",
                            "formControl": "MuiInputBase-formControl MuiInput-formControl",
                            "fullWidth": "MuiInputBase-fullWidth MuiInput-fullWidth",
                            "input": "MuiInputBase-input MuiInput-input makeStyles-inputInput-6",
                            "inputAdornedEnd": "MuiInputBase-inputAdornedEnd",
                            "inputAdornedStart": "MuiInputBase-inputAdornedStart",
                            "inputMarginDense": "MuiInputBase-inputMarginDense MuiInput-inputMarginDense",
                            "inputMultiline": "MuiInputBase-inputMultiline MuiInput-inputMultiline",
                            "inputSelect": "MuiInputBase-inputSelect",
                            "inputTypeSearch": "MuiInputBase-inputTypeSearch MuiInput-inputTypeSearch",
                            "marginDense": "MuiInputBase-marginDense",
                            "multiline": "MuiInputBase-multiline MuiInput-multiline",
                            "root": "MuiInputBase-root MuiInput-root makeStyles-inputRoot-5 MuiInput-underline",
                          }
                        }
                        fullWidth={true}
                        inputComponent="input"
                        inputProps={
                          Object {
                            "aria-activedescendant": null,
                            "aria-autocomplete": "list",
                            "aria-controls": null,
                            "aria-labelledby": "downshift-0-label",
                            "autoComplete": "off",
                            "id": "downshift-0-input",
                            "onKeyDown": [Function],
                            "placeholder": "Start typing",
                            "value": "",
                          }
                        }
                        multiline={false}
                        onBlur={[Function]}
                        onChange={[Function]}
                        onFocus={[Function]}
                        type="text"
                      >
                        <div
                          className="MuiInputBase-root MuiInput-root makeStyles-inputRoot-5 MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl"
                          onClick={[Function]}
                        >
                          <input
                            aria-activedescendant={null}
                            aria-autocomplete="list"
                            aria-controls={null}
                            aria-invalid={false}
                            aria-labelledby="downshift-0-label"
                            autoComplete="off"
                            className="MuiInputBase-input MuiInput-input makeStyles-inputInput-6"
                            disabled={false}
                            id="downshift-0-input"
                            onBlur={[Function]}
                            onChange={[Function]}
                            onFocus={[Function]}
                            onKeyDown={[Function]}
                            placeholder="Start typing"
                            required={false}
                            type="text"
                            value=""
                          />
                        </div>
                      </ForwardRef(InputBase)>
                    </WithStyles(ForwardRef(InputBase))>
                  </ForwardRef(Input)>
                </WithStyles(ForwardRef(Input))>
              </div>
            </ForwardRef(FormControl)>
          </WithStyles(ForwardRef(FormControl))>
        </ForwardRef(TextField)>
      </WithStyles(ForwardRef(TextField))>
      <div
        aria-labelledby="downshift-0-label"
        id="downshift-0-menu"
        role="listbox"
      />
    </div>
  </Downshift>
</Autocomplete>
`;
